File: app/src/main/java/com/github/nikololoshka/pepegaschedule/schedule/pair/DayOfWeek.java
package com.github.nikololoshka.pepegaschedule.schedule.pair;

import androidx.annotation.Nullable;

import com.github.nikololoshka.pepegaschedule.schedule.pair.exceptions.InvalidDayOfWeekException;

import java.util.Calendar;

public enum DayOfWeek {
    MONDAY,
    TUESDAY,
    WEDNESDAY,
    THURSDAY,
    FRIDAY,
    SATURDAY;

    static public DayOfWeek valueOf(@Nullable Calendar date) {
        if (date == null) {
            throw new IllegalArgumentException("Day of week is null");
        }

        int dayNumber = date.get(Calendar.DAY_OF_WEEK);
        switch (dayNumber) {
            case Calendar.MONDAY:
                return DayOfWeek.MONDAY;
            case Calendar.TUESDAY:
                return DayOfWeek.TUESDAY;
            case Calendar.WEDNESDAY:
                return DayOfWeek.WEDNESDAY;
            case Calendar.THURSDAY:
                return DayOfWeek.THURSDAY;
            case Calendar.FRIDAY:
                return DayOfWeek.FRIDAY;
            case Calendar.SATURDAY:
                return DayOfWeek.SATURDAY;
        }

        throw new InvalidDayOfWeekException(date);
    }
}

File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/model/schedule/pair/DayOfWeek.kt
package com.vereshchagin.nikolay.stankinschedule.model.schedule.pair

import android.os.Parcelable
import com.vereshchagin.nikolay.stankinschedule.model.schedule.DateDayOfWeekException
import kotlinx.parcelize.Parcelize
import org.joda.time.DateTimeConstants
import org.joda.time.LocalDate

/**
 * Дни недели даты в расписании.
 */
@Parcelize
enum class DayOfWeek : Parcelable {
    /**
     * Понедельник.
     */
    MONDAY,

    /**
     * Вторник.
     */
    TUESDAY,

    /**
     * Среда.
     */
    WEDNESDAY,

    /**
     * Четверг.
     */
    THURSDAY,

    /**
     * Пятница.
     */
    FRIDAY,

    /**
     * Суббота.
     */
    SATURDAY;

    companion object {
        /**
         * Возвращает день недели соответствующей даты.
         * @param date дата.
         * @return день недели.
         * @throws DateDayOfWeekException если не удалось узнать день недели.
         */
        fun of(date: LocalDate) : DayOfWeek {
            return when(date.dayOfWeek) {
                DateTimeConstants.MONDAY -> MONDAY
                DateTimeConstants.TUESDAY -> TUESDAY
                DateTimeConstants.WEDNESDAY -> WEDNESDAY
                DateTimeConstants.THURSDAY -> THURSDAY
                DateTimeConstants.FRIDAY -> FRIDAY
                DateTimeConstants.SATURDAY -> SATURDAY
                else -> {
                    throw DateDayOfWeekException( "Invalid day of week: $date")
                }
            }
        }
    }
}